File: leetcode_src/src/PartitionList.py
from typing import *
from LinkedListUtil import *

class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        beforeHead = ListNode(-1, None)
        afterHead = ListNode(-1, None)
        before = beforeHead
        after = afterHead

        while head:
            if head.val < x:
                before.next = head
                before = before.next
            else:
                after.next = head
                after = after.next

            head = head.next

        after.next = None
        before.next = afterHead.next

        return beforeHead.next


l1 = generateLinkedList("1,4,3,2,5,2")
l1 = Solution().partition(l1, 3)
l1.print()

File: leetcode_src/src/PartitionList.java
public class PartitionList {
    public ListNode partition(ListNode head, int x) {
        ListNode dummy = new ListNode(-1, head);
        ListNode pivot = dummy;
        ListNode prei = dummy;
        ListNode i = head;

        while (i != null) {
            if (i.val < x) {
                prei.next = i.next;
                i.next = pivot.next;
                pivot.next = i;

                if (prei.next != i) {
                    i = prei.next;
                }
                else {
                    prei = prei.next;
                    i = prei.next;
                }
                pivot = pivot.next;
            } else {
                prei = i;
                i = i.next;
            }
        }

        return dummy.next;
    }

    public static void main(String[] args) {
        PartitionList sol = new PartitionList();

        ListNode l1 = LinkedListUtil.generateLinkedList("1,4,3,2,5,2");
        sol.partition(l1, 4);
        l1.print();
    }
}
